src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,296 @@
+#include "paymentdisclosuredb.h"
+
+#include "zcash/Note.hpp"
+#include "zcash/NoteEncryption.hpp"
+
+using namespace std;
+using namespace libzcash;
+
+// Function declaration for function implemented in wallet/rpcwallet.cpp
+bool EnsureWalletIsAvailable(bool avoidException);
+
+/**
+ * RPC call to generate a payment disclosure
+ */
+UniValue z_getpaymentdisclosure(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    auto fEnablePaymentDisclosure = fExperimentalMode && GetBoolArg("-paymentdisclosure", false);
+    string strPaymentDisclosureDisabledMsg = "";
+    if (!fEnablePaymentDisclosure) {
+        strPaymentDisclosureDisabledMsg = "\nWARNING: Payment disclosure is currently DISABLED. This call always fails.\n";
+    }
+
+    if (fHelp || params.size() < 3 || params.size() > 4 )
+        throw runtime_error(
+            "z_getpaymentdisclosure \"txid\" \"js_index\" \"output_index\" (\"message\") \n"
+            "\nGenerate a payment disclosure for a given joinsplit output.\n"
+            "\nEXPERIMENTAL FEATURE\n"
+            + strPaymentDisclosureDisabledMsg +
+            "\nArguments:\n"
+            "1. \"txid\"            (string, required) \n"
+            "2. \"js_index\"        (string, required) \n"
+            "3. \"output_index\"    (string, required) \n"
+            "4. \"message\"         (string, optional) \n"
+            "\nResult:\n"
+            "\"paymentblob\"        (string) Hex string of payment blob\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_getpaymentdisclosure", "96f12882450429324d5f3b48630e3168220e49ab7b0f066e5c2935a6b88bb0f2 0 0 \"refund\"")
+            + HelpExampleRpc("z_getpaymentdisclosure", "\"96f12882450429324d5f3b48630e3168220e49ab7b0f066e5c2935a6b88bb0f2\", 0, 0, \"refund\"")
+        );
+
+    if (!fEnablePaymentDisclosure) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: payment disclosure is disabled.");
+    }
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+
+    // Check wallet knows about txid
+    string txid = params[0].get_str();
+    uint256 hash;
+    hash.SetHex(txid);
+
+    CTransaction tx;
+    uint256 hashBlock;
+
+    // Check txid has been seen
+    if (!GetTransaction(hash, tx, hashBlock, true)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available about transaction");
+    }
+
+    // Check tx has been confirmed
+    if (hashBlock.IsNull()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction has not been confirmed yet");
+    }
+
+    // Check is mine
+    if (!pwalletMain->mapWallet.count(hash)) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction does not belong to the wallet");
+    }
+    const CWalletTx& wtx = pwalletMain->mapWallet[hash];
+
+    // Check if shielded tx
+    if (wtx.vjoinsplit.empty()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction is not a shielded transaction");        
```

Comment:
Nit: we use "shielded transaction" colloquially to mean a fully-shielded transaction, vs. "shielding" and "unshielding", whereas this message should refer to all three. Not sure how to reword, but it should happen at some point.